### PR TITLE
Preserve memory format for torch.cat on CUDA

### LINKED ISSUE #34084: CUDA's `cat` implementation doesn't support memory_format
```python
x = torch.randn(200, 512, 28, 28, device='cuda', dtype=torch.float16).contiguous(memory_format=torch.channels_last)
y = torch.cat((x,x))
y.is_contiguous(memory_format=torch.channels_last) # False :(
```

cc @ngimel @VitalyFedyunin @jamesr66a

### PR DESCRIPTION
**Pull request by zasdfgbnm**
fix https://github.com/pytorch/pytorch/issues/34084

cc: @ptrblck @VitalyFedyunin

### CODE REVIEW

File: aten/src/ATen/native/cuda/Shape.cu [hunk 1]
```diff
@@ -159,9 +165,22 @@ void parallel_cat(Tensor &out, const TensorList &inputs, int64_t dimension,
...
   // Next, let's initialize the size, stride arrays for the output Tensor.
-  for (int i = 0; i < nDims; ++i) {
-    param.outputSize[i] = at::native::size(out, i);
-    param.outputStride[i] = out.stride(i);
+  if (memory_format == c10::MemoryFormat::Contiguous) {
+    for (int i = 0; i < nDims; ++i) {
+      param.outputSize[i] = at::native::size(out, i);
+      param.outputStride[i] = out.stride(i);
+    }
+  } else {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Shape.cu]: Can you have an `elseif` for channels-last memory format here, and TORCH_CHECK with "unsupported memory format" in the `else` branch.

File: aten/src/ATen/native/cuda/Shape.cu [hunk 2]
```diff
@@ -240,6 +270,27 @@ Tensor cat_cuda(TensorList inputs, int64_t dimension) {
...
+  for (auto &t : inputs) {
+    auto f = t.suggest_memory_format();
+    if (!format.has_value()) {
+      format = f;
+      continue;
+    }
+    if (format.value() == f) {
+      continue;
+    }
+    if (format.value() == c10::MemoryFormat::Contiguous || f == c10::MemoryFormat::Contiguous) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Shape.cu]: nit: this and the following conditional can be unified?

File: aten/src/ATen/native/cuda/Shape.cu [hunk 3]
```diff
@@ -240,6 +270,27 @@ Tensor cat_cuda(TensorList inputs, int64_t dimension) {
   return out;
 }
 
+inline c10::MemoryFormat compute_output_memory_format(const TensorList &inputs) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Shape.cu]: can you do it with std::all_of?

File: aten/src/ATen/native/cuda/Shape.cu [hunk 4]
```diff
@@ -240,6 +270,27 @@ Tensor cat_cuda(TensorList inputs, int64_t dimension) {
...
+  c10::optional<c10::MemoryFormat> format = c10::nullopt;
+  for (auto &t : inputs) {
+    auto f = t.suggest_memory_format();
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Shape.cu]: what happens if one of the tensors is undefined?

File: aten/src/ATen/native/cuda/Shape.cu (hunk 1 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Shape.cu]: fixed

File: aten/src/ATen/native/cuda/Shape.cu (hunk 2 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Shape.cu]: fixed

File: aten/src/ATen/native/cuda/Shape.cu (hunk 4 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Shape.cu]: Do we need to check if tensors are defined? There is no such check in shape computation.

Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Shape.cu]: I don't think we should do defined check in this function, but we might need to do it in `cat_out`.

File: aten/src/ATen/native/cuda/Shape.cu (hunk 3 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Shape.cu]: I don't think this is possible. `std::all_of` is for boolean values, but memory format has `ChannelsLast` and `ChannelsLast3d`, we need to preserve both case.

File: aten/src/ATen/native/cuda/Shape.cu (hunk 4 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/Shape.cu]: I don't really know if it's needed here, I've seen it below in `allContiguous`. Would `suggest_memory_format` even work on undefined tensor?

File: aten/src/ATen/native/cuda/Shape.cu (hunk 3 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/Shape.cu]: Cool, I forgot about `ChannelsLast3d`.

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/cuda/Shape.cu (hunk 4 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/Shape.cu]: No, it won't. It calls `sizes()` which is checked for undefined tensor.

**Comment by zasdfgbnm** [comment]
@ngimel Just resolved a conflict, you might need to import again.

**Pull request closed without merging**